Fix #3164

jwiesler opened the pull request.

mattulbrich reviewed (changes requested):
During review, the following occurred:
* I loaded SumAndMax and selected arbitrary nodes in the proof tree
* Tooltips were shortened (also the active statement?)
* Suddenly, I could no longer select nodes in the tree.
* The following stacktrace appeared on the console:
* (I may have double clicked on a node??)
```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException: Cannot invoke "javax.swing.tree.TreeCellEditor.stopCellEditing()" because "this.cellEditor" is null
        at java.desktop/javax.swing.plaf.basic.BasicTreeUI.stopEditing(BasicTreeUI.java:748)
        at java.desktop/javax.swing.plaf.basic.BasicTreeUI$Handler.mousePressed(BasicTreeUI.java:4026)
        at java.desktop/java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:288)
        at java.desktop/java.awt.Component.processMouseEvent(Component.java:6623)
        at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3389)
        at java.desktop/java.awt.Component.processEvent(Component.java:6391)
        at java.desktop/java.awt.Container.processEvent(Container.java:2266)
        at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5001)
        at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
        at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
        at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4948)
        at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4572)
        at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4516)
        at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
        at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2780)
        at java.desktop/java.awt.Component.dispatchEvent(Component.java:4833)
        at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
        at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
        at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:746)
        at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:744)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
        at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:743)
        at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
        at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
        at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

jwiesler commented:
> (I may have double clicked on a node??)

This bug has been happening for months and I already mentioned it somewhere, I think it is a bug in swing or the code that handles the editor.
It happens when you expand a node by double clicking on it (maybe 3 times?). This starts editing the branch name and this editing can't be closed by any means.

jwiesler commented:
Just as a note: I think this is the same issue that the example dialog tree and other trees had (and I fixed). I found a way to reproduce: 
* SumAndMax
* Preparation only
* Scroll all the way down
* *Triple* click to collapse the node `Invariant Initially Valid`
* Scrolling up occurs because the tree got shorter
* Make sure to position the mouse x-wise such that it will land on a node (not too far to the right)
* This node is now being edited and the bug appears

jwiesler commented:
Ok I fixed the issue.

I also added a max length to the active statement.

This is all best effort since both active statement and applied on can potentially become large and I don't know how to distribute that space between the two. So for now active statement <= 5 lines and applied on <= whatever is set in the gui.

mattulbrich commented on key.util/src/main/java/org/key_project/util/java/StringUtil.java:
```diff
@@ -433,4 +433,24 @@ public static String trim(String text, String chars) {
+     * @param ch what to look for
+     * @return the number of occurrences
+     */
+    public static int countMatches(String str, char ch) {
```
I wonder if `countMatches` is the right name for the method.

When googling for a better name, I found that this code is taken from Apache commons.
That must be attributed. We probably have to mention this somewhere.

mattulbrich replied:
/cc @wadoon. Do we have some license automation in the workflows?

jwiesler replied:
Yes that was taken from Apache Commons and should be named `countOccurences`. Do we really need to attribute this? It's 5 LOC that would have looked exactly the same if I just typed them by hand.

wadoon replied:
There are plugins to check for license violatons, see https://github.com/jk1/Gradle-License-Report, but this does not help in these case, as this is not a Gradle dependency. 

In this case, a note in the Javadoc should be enough to give credits to Apache Commons. Apache License should be compatible with GPLv2.

jwiesler replied:
Removed the function since it was unused. I think we should just include apache commons, since it has a lot of useful functions.

mattulbrich commented:
> Just as a note: I think this is the same issue that the example dialog tree and other trees had (and I fixed). I found a way to reproduce

Nice catch!

I can still triple click on branch labels and change them. Is that a feature?

jwiesler commented:
> I can still triple click on branch labels and change them. Is that a feature?

Yep, that was always intended. I think they are even saved (didn't test).

jwiesler commented:
@WolframPfeifer added the `PositionInfo` cleanup we just talked about here as well.

WolframPfeifer commented on key.ui/src/main/java/de/uka/ilkd/key/gui/prooftree/ProofTreeView.java:
```diff
@@ -192,6 +193,30 @@ public void updateUI() {
+        var renderer = delegateView.getCellRenderer() instanceof DefaultTreeCellRenderer
+                ? (DefaultTreeCellRenderer) delegateView.getCellRenderer()
+                : null;
+        delegateView.setCellEditor(new DefaultTreeCellEditor(delegateView, renderer) {
```
Ideally, there should be a comment here about the purpose of the anonymous class (it is there to disable editing on non-branching nodes which are not the root, right?).

WolframPfeifer reviewed (approved):
The tooltip changes work nicely. It is a bit strange that you still can get into edit mode on every node by pressing "F2". However, the changes are only applied (when you confirm the edit operation by pressing "Enter") when you are editing the text of a branch node ...

Since the PR definitely improves upon the status quo, I am ok with that. It would be nice if there was a comment about the purpose about the anonymous editor class.

WolframPfeifer commented:
> > I can still triple click on branch labels and change them. Is that a feature?
> 
> Yep, that was always intended. I think they are even saved (didn't test).

I tested it: Indeed, you can set a custom text for branch labels, which is then saved in the proof file.
